### PR TITLE
feat: Part 3 - Introduce Vector Store Manager and Vector Store Class

### PR DESCRIPTION
**Pull request by ishaansehgal99**
**Reason for Change**:
This series of PR will integrate llamaindex RAG service for Kaito.

This PR introduces the main logic for indexing, querying, adding and listing documents.`faiss_store.py` is the most important file. 

`faiss_store.py` introduces the code to perform RAG operations, including persisting and loading DB from disk incase it is needed. PR also includes tests for `faiss_store.py` in `test_faiss_store.py`. `conftest.py` is a test configuration file, in it we ensure the tests run synchronously and CPU only.

`manager.py` is a light abstraction around the API (`main.py`) which will be introduced in subsequent PRs.

### CODE REVIEW

File: ragengine/vector_store/faiss_store.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+import faiss
+from llama_index.core import Document as LlamaDocument
+from llama_index.core import (StorageContext, VectorStoreIndex, load_index_from_storage)
+from llama_index.core.storage.index_store import SimpleIndexStore
+from llama_index.core.storage.docstore.types import RefDocInfo
+from llama_index.vector_stores.faiss import FaissVectorStore
+
+from models import Document
+from inference.custom_inference import CustomInference
```
Comment (zhuangqh) [inline ragengine/vector_store/faiss_store.py:12]: package name and class name have been changed.
```suggestion
from inference.inference import Inference
```

File: ragengine/vector_store/faiss_store.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+from config import PERSIST_DIR
+
+from .base import BaseVectorStore
+
+
+class FaissVectorStoreHandler(BaseVectorStore):
+    def __init__(self, embedding_manager):
```
Comment (zhuangqh) [inline ragengine/vector_store/faiss_store.py:20]: add type assertion here.
embedding_manager: BaseEmbeddingModel

File: ragengine/vector_store/faiss_store.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+from llama_index.core import Document as LlamaDocument
+from llama_index.core import (StorageContext, VectorStoreIndex, load_index_from_storage)
+from llama_index.core.storage.index_store import SimpleIndexStore
+from llama_index.core.storage.docstore.types import RefDocInfo
+from llama_index.vector_stores.faiss import FaissVectorStore
+
+from models import Document
+from inference.custom_inference import CustomInference
+
+from config import PERSIST_DIR
```
Comment (zhuangqh) [inline ragengine/vector_store/faiss_store.py:14]: if import pkg in this way, can python find the correct local pkg? not some python built-in library.
I think the correct way may be `from ragengine.config import XXXX`

File: ragengine/vector_store/faiss_store.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+from typing import Dict, List
+
+import faiss
+from llama_index.core import Document as LlamaDocument
+from llama_index.core import (StorageContext, VectorStoreIndex, load_index_from_storage)
+from llama_index.core.storage.index_store import SimpleIndexStore
+from llama_index.core.storage.docstore.types import RefDocInfo
+from llama_index.vector_stores.faiss import FaissVectorStore
+
+from models import Document
```
Comment (zhuangqh) [inline ragengine/vector_store/faiss_store.py:11]: It looks like `models.py` was forgotten to be added to this pr.

File: ragengine/tests/vector_store/test_faiss_store.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    assert vector_store_manager.document_exists("test_index", "1")
+    assert vector_store_manager.document_exists("another_test_index", "1")
+
+    vector_store_manager.index_map = {}  # Clear current in-memory index map
+    loaded_index = vector_store_manager._load_index("test_index")
+
+    assert loaded_index is not None
+    assert vector_store_manager.document_exists("test_index", "1")
+    assert not vector_store_manager.document_exists("another_test_index", "1") # Since we didn't load this index
+
```
Comment (zhuangqh) [inline ragengine/tests/vector_store/test_faiss_store.py:225]: how to run this test

File: ragengine/tests/vector_store/test_faiss_store.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    assert query_result.response == "This is the completion from the API"
+
+    mock_post.assert_called_once_with(
+        INFERENCE_URL,
+        # Auto-Generated by LlamaIndex
+        json={"prompt": "Context information is below.\n---------------------\ntype: text\n\nFirst document\n---------------------\nGiven the context information and not prior knowledge, answer the query.\nQuery: First\nAnswer: ", "formatted": True, 'temperature': 0.7},
+        headers={"Authorization": f"Bearer {INFERENCE_ACCESS_SECRET}"}
+    )
+
+"""
```
Comment (Fei-Guo) [inline ragengine/tests/vector_store/test_faiss_store.py:89]: Let us remove all the commented code. They serve no values for now.

File: ragengine/vector_store/base.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+    @abstractmethod
+    def query(self, index_name: str, query: str, top_k: int, params: dict):
+        pass
+
+    @abstractmethod
+    def add_document(self, index_name: str, document: Document):
+        pass
+
+    """
```
Comment (Fei-Guo) [inline ragengine/vector_store/base.py:21]: Remove commented codes.

File: ragengine/vector_store_manager/manager.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    def create(self, index_name: str, documents: List[Document]) -> List[str]:
+        """Index new documents."""
+        return self.vector_store.index_documents(index_name, documents)
+
+    def read(self, index_name: str, query: str, top_k: int, llm_params: dict):
+        """Query the indexed documents."""
+        return self.vector_store.query(index_name, query, top_k, llm_params)
+
+    """
```
Comment (Fei-Guo) [inline ragengine/vector_store_manager/manager.py:20]: remove commented code.

File: ragengine/vector_store/faiss_store.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    def get_document(self, index_name: str, doc_id: str) -> RefDocInfo:
+        """Retrieves a document's RefDocInfo by its ID."""
+        if index_name not in self.index_map:
+            raise ValueError(f"No such index: '{index_name}' exists.")
+
+        # Try to retrieve the RefDocInfo associated with the doc_id
+        ref_doc_info = self.index_map[index_name].ref_doc_info.get(doc_id)
+        return ref_doc_info
+
+    """
```
Comment (Fei-Guo) [inline ragengine/vector_store/faiss_store.py:100]: remove commented code.

File: ragengine/vector_store/faiss_store.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        """Loads the global SimpleIndexStore from disk."""
+        store_path = os.path.join(PERSIST_DIR, "store.json")
+
+        if not os.path.exists(store_path):
+            raise ValueError("No persisted index store found.")
+
+        # Load the global index store from the persisted JSON
+        self.index_store = SimpleIndexStore.from_persist_path(store_path)
+
+    def _load_indices(self):
```
Comment (Fei-Guo) [inline ragengine/vector_store/faiss_store.py:172]: If we don't plan to support loading index from local disk in this release, remove the related code.

File: ragengine/vector_store/faiss_store.py (hunk 4 above)
Comment (ishaansehgal99) [reply ragengine/vector_store/faiss_store.py:11]: Added

File: ragengine/tests/vector_store/test_faiss_store.py (hunk 6 above)
Comment (ishaansehgal99) [reply ragengine/tests/vector_store/test_faiss_store.py]: done

File: ragengine/vector_store/base.py (hunk 7 above)
Comment (ishaansehgal99) [reply ragengine/vector_store/base.py]: done

File: ragengine/vector_store/faiss_store.py (hunk 9 above)
Comment (ishaansehgal99) [reply ragengine/vector_store/faiss_store.py]: done

File: ragengine/vector_store_manager/manager.py (hunk 8 above)
Comment (ishaansehgal99) [reply ragengine/vector_store_manager/manager.py]: done

File: ragengine/vector_store/faiss_store.py (hunk 10 above)
Comment (ishaansehgal99) [reply ragengine/vector_store/faiss_store.py:112]: Currently persist is the only function called every "index" call. It just saves db to a file, load is never called

Comment (ishaansehgal99) [reply ragengine/vector_store/faiss_store.py:112]: removed load code

File: ragengine/tests/vector_store/test_faiss_store.py (hunk 5 above)
Comment (ishaansehgal99) [reply ragengine/tests/vector_store/test_faiss_store.py:142]: `pytest -o log_cli=true -o log_cli_level=INFO ragengine/tests`
next PR updates test.yml (unit-tests) and makefile to include tests

File: ragengine/tests/vector_store/test_faiss_store.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+from config import MODEL_ID, INFERENCE_URL, INFERENCE_ACCESS_SECRET
+
+@pytest.fixture(scope='session')
+def init_embed_manager():
+    return LocalHuggingFaceEmbedding(MODEL_ID)
+
+@pytest.fixture
+def vector_store_manager(init_embed_manager):
+    with TemporaryDirectory() as temp_dir:
+        print(f"Saving Temporary Test Storage at: {temp_dir}")
```
[the remark this answers is not in the document]
Comment (ishaansehgal99) [reply ragengine/tests/vector_store/test_faiss_store.py:18]: sure

File: ragengine/vector_store/faiss_store.py (hunk 2 above)
Comment (ishaansehgal99) [reply ragengine/vector_store/faiss_store.py]: done

File: ragengine/vector_store/faiss_store.py (hunk 3 above)
Comment (ishaansehgal99) [reply ragengine/vector_store/faiss_store.py:14]: right, I was editing the syspath in conftest.py using absolute paths is better, done

**Review by zhuangqh — APPROVED** [review]

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**